---
title: "`r gsub('.Rmd', ' ', gsub('_', ' ', knitr::current_input()))`"
author: "`r Sys.getenv('USER')`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
---

```{r setup}

library(tidyverse)
library(dplyr)
library(tidyr)
library(janitor)

```

## Summarize results of bulk processing for outlier lead accuracy project

This data comes from the CKCC1 outlier results.

## Identify input data

```{r}

data_release_id <- "CKCC_outlier_lead_accuracy.results.2020_07_17"

```

# Load data
```{r}

gathered_results <- read_tsv(paste0("bulk_results.", data_release_id, ".tsv.gz")) %>%
  rename(Mean_Accuracy = `Mean Accuracy`)

```


# Add bin info
```{r}

exp_bins <- c(0, 1, 3, 5, 7, 10)
expression_bins_all <- c("0-1", "1-3", "3-5", "5-7", "7-10", "10+")

# length_bins <- c(0, 200, 400, 600, 800, 1000, 1200, 1500, 2000, 3000, 4000)
# length_bins_all <- c("0-200",
#                      "200-400",
#                      "400-600",
#                      "600-800",
#                      "800-1000",
#                      "1000-1200",
#                      "1200-1500",
#                      "1500-2000",
#                      "2000-3000",
#                      "3000-4000",
#                      "4000+")

length_bins <- seq(0, 3000, 300)
length_bins_all <- paste(length_bins, lead(length_bins), sep = "-") %>%
  str_replace("-NA", "+")


bulk_results <- gathered_results %>%
  mutate(MEND_bin = round((MEND_depth/1e6) / 4) * 4,
         MEND_bin = ifelse(MEND_depth>44*1e6, 44, MEND_bin),
         Expression_bin = expression_bins_all[findInterval(Expression, exp_bins)],
         length_bin = length_bins_all[findInterval(effective_length, length_bins)])

table(bulk_results$MEND_bin)
```

# Confirm the absence of NA values
```{r}

dim(bulk_results)
dim(is.na(bulk_results))

```

## Profile of MEND depths
```{r}
just_MEND_depths <- bulk_results %>% 
  select(Sample, MEND_depth)

ggplot(just_MEND_depths) +
  geom_histogram(aes(x=MEND_depth/1E6))

range(just_MEND_depths$MEND_depth/1e6)  

range(summary(just_MEND_depths$MEND_depth/1e6))  

```


## Number of outlier leads per sample

```{r}

# Histogram of number of outlier leads per sample (split by outlier type & fill is depth)

num_leads <- bulk_results %>% group_by(Sample, Type, MEND_bin) %>% tally()

plot1 <- ggplot(num_leads, aes(x=n, fill=as.factor(MEND_bin))) +
  geom_histogram() +
  facet_wrap(~Type) +
  ggtitle("Number of outlier leads per sample")

plot1

```

## Look at bin sizes

```{r}

# Make sure (depth, expression, length) bins are relatively similar sizes

all_bins <- bulk_results %>% tabyl(MEND_bin, Expression_bin, length_bin)

all_bins

```

## Spectrum of accuracies by depth, expression, and length

```{r}

plot2 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~MEND_bin) +
  ggtitle("Spectrum of accuracies by MEND depth")

plot2

plot3 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~Expression_bin) +
  ggtitle("Spectrum of accuracies by expression level")

plot3

plot4 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~length_bin) +
  ggtitle("Spectrum of accuracies by effective gene length")

plot4

```

# Spectrum of accuracies by outlier type (histograms for figure 1)

```{r}

bulk_results$Type <- factor(bulk_results$Type, levels = c("Up", "Down"))

ggplot(bulk_results, aes(x=Mean_Accuracy, color=Type)) +
  geom_histogram(fill="white") +
  facet_wrap(~Type, scales="free") +
  theme_bw() +
  xlab("Accuracy") +
  ylab("Gene expression measurements") +
  ggtitle("Accuracies of up and down outlier gene expression measurements")

ggsave("../figures/accuracies_all.png")

```

# Distributions of accuracies for different expressions, depths, and lengths (boxplots for figure 1)

```{r}

figure1_data <- bulk_results %>%
  mutate(exp_gt5 = Expression >= 5,
         depth_gt20 = MEND_bin >= 20,
         length_gt1000 = effective_length >= 1000)

# Expression above or below 5

ggplot(subset(figure1_data, Type == "Up"), aes(x=exp_gt5, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Exp >= 5", "FALSE"="Exp < 5")) +
  ylab("Accuracy") +
  xlab("Expression above or below 5 log2(TPM+1)") +
  ggtitle("Accuracies of up outliers with high and low gene expression")

ggsave("../figures/accuracies_up_exp.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=exp_gt5, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Exp >= 5", "FALSE"="Exp < 5")) +
  ylab("Accuracy") +
  xlab("Gene expression above or below 5 log2(TPM+1)") +
  ggtitle("Accuracies of down outliers with high and low gene expression")

ggsave("../figures/accuracies_down_exp.png")

# Depth above or below 20

ggplot(subset(figure1_data, Type == "Up"), aes(x=depth_gt20, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Depth >= 20", "FALSE"="Depth < 20")) +
  ylab("Accuracy") +
  xlab("MEND depth above or below 20 million MEND reads") +
  ggtitle("Accuracies of up outliers with high and low MEND read depth")

ggsave("../figures/accuracies_up_depth.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=depth_gt20, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Depth >= 20", "FALSE"="Depth < 20")) +
  ylab("Accuracy") +
  xlab("MEND depth above or below 20 million MEND reads") +
  ggtitle("Accuracies of down outliers with high and low MEND read depth")

ggsave("../figures/accuracies_down_depth.png")

# Length above or below 1000

ggplot(subset(figure1_data, Type == "Up"), aes(x=length_gt1000, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Length >= 1000", "FALSE"="Length < 1000")) +
  ylab("Accuracy") +
  xlab("Gene length above or below 1000 base pairs") +
  ggtitle("Accuracies of up outliers with high and low effective gene length")

ggsave("../figures/accuracies_up_length.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=length_gt1000, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Length >= 1000", "FALSE"="Length < 1000")) +
  ylab("Accuracy") +
  xlab("Gene length above or below 1000 base pairs") +
  ggtitle("Accuracies of down outliers with high and low effective gene length")

ggsave("../figures/accuracies_down_length.png")

```

Random calculations for the paper

```{r}

# Average number of up and down outliers per sample

mean(subset(num_leads, Type == "Up")$n)
mean(subset(num_leads, Type == "Down")$n)

median(subset(num_leads, Type == "Up")$n)
median(subset(num_leads, Type == "Down")$n)

# Average accuracies of up and down outliers per sample

mean(subset(figure1_data, Type == "Up" & !is.na(Mean_Accuracy))$Mean_Accuracy)
mean(subset(figure1_data, Type == "Down" & !is.na(Mean_Accuracy))$Mean_Accuracy)

median(subset(figure1_data, Type == "Up" & !is.na(Mean_Accuracy))$Mean_Accuracy)
median(subset(figure1_data, Type == "Down" & !is.na(Mean_Accuracy))$Mean_Accuracy)

# Fraction of accuracies above 95%

count(subset(figure1_data, Type == "Up" & Mean_Accuracy>95)) /
  count(subset(figure1_data, Type == "Up"))

count(subset(figure1_data, Type == "Down" & Mean_Accuracy>95)) /
  count(subset(figure1_data, Type == "Down"))

count(subset(figure1_data, Mean_Accuracy>95)) /
  count(figure1_data)

# the bump in distribution due to the 0-1 crowd

median(subset(figure1_data, Expression_bin == "0-1")$Mean_Accuracy)

```

