---
title: "`r gsub('.Rmd', ' ', gsub('_', ' ', knitr::current_input()))`"
author: "`r Sys.getenv('USER')`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
---

```{r setup, echo=FALSE}

library(tidyverse)
library(dplyr)
library(tidyr)
library(janitor)

```

## Summarize results of bulk processing for outlier lead accuracy project

This data comes from the CKCC1 outlier results.

## Load data and create summary tsv (only run once)

```{r, echo=FALSE}

samples <- list.files(path="/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/CKCC_outlier_lead_accuracy.results.2020_07_17/")
first <- TRUE

for (s in samples) {
  curr_wd <- paste0("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/CKCC_outlier_lead_accuracy.results.2020_07_17/", s, "/")
  setwd(curr_wd)
  leads <- list.files(path=curr_wd) %>%
    lapply(read_tsv) %>%
    bind_rows
  if (first == TRUE) {
    results_all <- leads
    first <- FALSE
  } else {
    results_all <- bind_rows(results_all, leads)
  }
}

# start of scratch to deal with T gene mess

results_all <- read_tsv("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/old-results_all.tsv")

# change T to TT and then unchange in tsv file

problem1 <- read_tsv("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/CKCC_outlier_lead_accuracy.results.2020_07_17/TH03_0012_S02/sample_TH03_0012_S02_gene_T_summary.tsv", col_types = list(Gene = col_character())) %>% type_convert

problem1$Gene <- "T"

# change T to TT and then unchange in tsv file

problem2 <- read_tsv("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/CKCC_outlier_lead_accuracy.results.2020_07_17/TH03_0146_S01/sample_TH03_0146_S01_gene_T_summary.tsv", col_types = list(Gene = col_character())) %>% type_convert

problem2$Gene <- "T"

# delete T gene summary file from directory and then undelete after this code chunk

s1 <- "TH03_0012_S02"
first <- TRUE
curr_wd <- paste0("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/CKCC_outlier_lead_accuracy.results.2020_07_17/", s1, "/")
setwd(curr_wd)
lead_files <- list.files(path=curr_wd)
for (l in lead_files) {
  print(l)
  linfo <- read_tsv(l) %>% type_convert
  if (first == TRUE) {
    leads <- linfo
    first <- FALSE
  } else {
    leads <- bind_rows(leads, linfo)
  }
}
leads1 <- leads

# delete T gene summary file from directory and then undelete after this code chunk

s2 <- "TH03_0146_S01"
first <- TRUE
curr_wd <- paste0("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/CKCC_outlier_lead_accuracy.results.2020_07_17/", s2, "/")
setwd(curr_wd)
lead_files <- list.files(path=curr_wd)
for (l in lead_files) {
  print(l)
  linfo <- read_tsv(l) %>% type_convert
  if (first == TRUE) {
    leads <- linfo
    first <- FALSE
  } else {
    leads <- bind_rows(leads, linfo)
  }
}
leads2 <- leads
  
test <- bind_rows(results_all, problem1, problem2, leads1, leads2)

# this worked so now this is the new results_all

results_all <- test

write_tsv(results_all, paste0("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results_all.tsv"))

# end of scratch

names(results_all) <- gsub(" ", "_", names(results_all))

# Add depth data

input_data <- read_tsv("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results/outlier_probability_input_CKCC_2020_07_13_batch1.tsv")
depth_data <- unique(input_data %>% select(sample_id, MEND))
depth_data <- rename(depth_data, c("Sample"="sample_id", "MEND_depth"="MEND"))
length_data <- input_data %>% select(sample_id, Gene, effective_length)
length_data <- rename(length_data, c("Sample"="sample_id"))
results_all_with_depth_and_length <- merge(results_all, depth_data) %>% merge(length_data)

# Add bin info

exp_bins <- c(0, 1, 3, 5, 7, 10)
expression_bins_all <- c("0-1", "1-3", "3-5", "5-7", "7-10", "10+")

length_bins <- c(0, 200, 400, 600, 800, 1000, 1200, 1500, 2000, 3000, 4000)
length_bins_all <- c("0-200",
                     "200-400",
                     "400-600",
                     "600-800",
                     "800-1000",
                     "1000-1200",
                     "1200-1500",
                     "1500-2000",
                     "2000-3000",
                     "3000-4000",
                     "4000+")

results_all_long <- results_all_with_depth_and_length %>%
  mutate(MEND_bin = round((MEND_depth/1e6) / 4) * 4,
         MEND_bin = ifelse(MEND_depth>44*1e6, 44, MEND_bin),
         Expression_bin = expression_bins_all[findInterval(Expression, exp_bins)],
         length_bin = length_bins_all[findInterval(effective_length, length_bins)])

# Save data to file to load it in the future

write_tsv(results_all_long, paste0("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results_all_long.tsv"))

```

## Load summary tsv

```{r, echo=FALSE}

bulk_results <- read_tsv("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/bulk_processing/results_all_long.tsv", col_types=cols())

bulk_results <- bulk_results[-c(8:14)]

```

## Number of outlier leads per sample

```{r, echo=FALSE}

# Histogram of number of outlier leads per sample (split by outlier type & fill is depth)

num_leads <- bulk_results %>% group_by(Sample, Type, MEND_bin) %>% tally()

plot1 <- ggplot(num_leads, aes(x=n, fill=as.factor(MEND_bin))) +
  geom_histogram() +
  facet_wrap(~Type) +
  ggtitle("Number of outlier leads per sample")

plot1

```

## Look at bin sizes

```{r, echo=FALSE}

# Make sure (depth, expression, length) bins are relatively similar sizes

all_bins <- bulk_results %>% tabyl(MEND_bin, Expression_bin, length_bin)

all_bins

```

## Spectrum of accuracies by depth, expression, and length

```{r, echo=FALSE}

plot2 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~MEND_bin) +
  ggtitle("Spectrum of accuracies by MEND depth")

plot2

plot3 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~Expression_bin) +
  ggtitle("Spectrum of accuracies by expression level")

plot3

plot4 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~length_bin) +
  ggtitle("Spectrum of accuracies by effective gene length")

plot4

```

# Spectrum of accuracies by outlier type (histograms for figure 1)

```{r, echo=FALSE}

bulk_results$Type <- factor(bulk_results$Type, levels = c("Up", "Down"))

ggplot(bulk_results, aes(x=Mean_Accuracy, color=Type)) +
  geom_histogram(fill="white") +
  facet_wrap(~Type, scales="free") +
  theme_bw() +
  xlab("Accuracy") +
  ylab("Gene expression measurements") +
  ggtitle("Accuracies of up and down outlier gene expression measurements")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_all.png")

```

# Distributions of accuracies for different expressions, depths, and lengths (boxplots for figure 1)

```{r, echo=FALSE}

figure1_data <- bulk_results %>%
  mutate(exp_gt5 = Expression >= 5,
         depth_gt20 = MEND_bin >= 20,
         length_gt1000 = effective_length >= 1000)

# Expression above or below 5

ggplot(subset(figure1_data, Type == "Up"), aes(x=exp_gt5, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Exp >= 5", "FALSE"="Exp < 5")) +
  ylab("Accuracy") +
  xlab("Expression above or below 5 log2(TPM+1)") +
  ggtitle("Accuracies of up outliers with high and low expression")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_up_exp.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=exp_gt5, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Exp >= 5", "FALSE"="Exp < 5")) +
  ylab("Accuracy") +
  xlab("Gene expression above or below 5 log2(TPM+1)") +
  ggtitle("Accuracies of down outliers with high and low gene expression")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_down_exp.png")

# Depth above or below 20

ggplot(subset(figure1_data, Type == "Up"), aes(x=depth_gt20, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Depth >= 20", "FALSE"="Depth < 20")) +
  ylab("Accuracy") +
  xlab("MEND depth above or below 20 million MEND reads") +
  ggtitle("Accuracies of up outliers with high and low MEND depth")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_up_depth.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=depth_gt20, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Depth >= 20", "FALSE"="Depth < 20")) +
  ylab("Accuracy") +
  xlab("MEND depth above or below 20 million MEND reads") +
  ggtitle("Accuracies of down outliers with high and low MEND depth")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_down_depth.png")

# Length above or below 1000

ggplot(subset(figure1_data, Type == "Up"), aes(x=length_gt1000, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Length >= 1000", "FALSE"="Length < 1000")) +
  ylab("Accuracy") +
  xlab("Gene length above or below 1000 base pairs") +
  ggtitle("Accuracies of up outliers with high and low gene length")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_up_length.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=length_gt1000, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Length >= 1000", "FALSE"="Length < 1000")) +
  ylab("Accuracy") +
  xlab("Gene length above or below 1000 base pairs") +
  ggtitle("Accuracies of down outliers with high and low gene length")

ggsave("/Users/jacquelynroger/Documents/research/treehouse/projects/gene_expression_outlier_accuracy/figures/accuracies_down_length.png")

```

